# Workbook

Use this notebook to complete the exercises throughout the workshop.

#### Table of Contents
- [Section 1 &ndash; Getting Started with Pandas](#Section-1)
- [Section 2 &ndash; Data Wrangling](#Section-2)
- [Section 3 &ndash; Data Visualization](#Section-3)

--- 

### Section 1

#### Exercise 1.1
##### Create a DataFrame by reading in the `2019_Yellow_Taxi_Trip_Data.csv` file. Examine the first 5 rows.

In [1]:
import pandas as pd

In [2]:

df = pd.read_csv('../data/2019_Yellow_Taxi_Trip_Data.csv')
df.head(5)

vendorid     tpep_pickup_datetime    tpep_dropoff_datetime  \
0         2  2019-10-23T16:39:42.000  2019-10-23T17:14:10.000   
1         1  2019-10-23T16:32:08.000  2019-10-23T16:45:26.000   
2         2  2019-10-23T16:08:44.000  2019-10-23T16:21:11.000   
3         2  2019-10-23T16:22:44.000  2019-10-23T16:43:26.000   
4         2  2019-10-23T16:45:11.000  2019-10-23T16:58:49.000   

   passenger_count  trip_distance  ratecodeid store_and_fwd_flag  \
0                1           7.93           1                  N   
1                1           2.00           1                  N   
2                1           1.36           1                  N   
3                1           1.00           1                  N   
4                1           1.96           1                  N   

   pulocationid  dolocationid  payment_type  fare_amount  extra  mta_tax  \
0           138           170             1         29.5    1.0      0.5   
1            11            26             1         10.5    1.0      0.5   
2           163           162             1          9.5    1.0      0.5   
3           170           163             1         13.0    1.0      0.5   
4           163           236             1         10.5    1.0      0.5   

   tip_amount  tolls_amount  improvement_surcharge  total_amount  \
0        7.98          6.12                    0.3         47.90   
1        0.00          0.00                    0.3         12.30   
2        2.00          0.00                    0.3         15.80   
3        4.32          0.00                    0.3         21.62   
4        0.50          0.00                    0.3         15.30   

   congestion_surcharge  
0                   2.5  
1                   0.0  
2                   2.5  
3                   2.5  
4                   2.5

#### Exercise 1.2
##### Find the dimensions (number of rows and number of columns) in the data.

In [3]:
df.shape

(10000, 18)

#### Exercise 1.3
##### Using the data in the `2019_Yellow_Taxi_Trip_Data.csv` file, calculate summary statistics for the `fare_amount`, `tip_amount`, `tolls_amount`, and `total_amount` columns.

In [4]:
df.columns

Index(['vendorid', 'tpep_pickup_datetime', 'tpep_dropoff_datetime',
       'passenger_count', 'trip_distance', 'ratecodeid', 'store_and_fwd_flag',
       'pulocationid', 'dolocationid', 'payment_type', 'fare_amount', 'extra',
       'mta_tax', 'tip_amount', 'tolls_amount', 'improvement_surcharge',
       'total_amount', 'congestion_surcharge'],
      dtype='object')

In [5]:
# Using the data in the `2019_Yellow_Taxi_Trip_Data.csv` file, calculate summary statistics for the `fare_amount`, `tip_amount`, `tolls_amount`, and `total_amount` columns.
df[['fare_amount', 'tip_amount', 'tolls_amount', 'total_amount']].describe()

fare_amount    tip_amount  tolls_amount  total_amount
count  10000.000000  10000.000000  10000.000000  10000.000000
mean      15.106313      2.634494      0.623447     22.564659
std       13.954762      3.409800      6.437507     19.209255
min      -52.000000      0.000000     -6.120000    -65.920000
25%        7.000000      0.000000      0.000000     12.375000
50%       10.000000      2.000000      0.000000     16.300000
75%       16.000000      3.250000      0.000000     22.880000
max      176.000000     43.000000    612.000000    671.800000

In [6]:
# you can also transpose the data
df[['fare_amount', 'tip_amount', 'tolls_amount', 'total_amount']].describe().T
# also you can save the data to a variable
# df_stats = df[['fare_amount', 'tip_amount', 'tolls_amount', 'total_amount']].describe().T
# you can also save the result to csv file
# df[['fare_amount', 'tip_amount', 'tolls_amount', 'total_amount']].describe().T.to_csv('summary.csv')

count       mean        std    min     25%   50%    75%    max
fare_amount   10000.0  15.106313  13.954762 -52.00   7.000  10.0  16.00  176.0
tip_amount    10000.0   2.634494   3.409800   0.00   0.000   2.0   3.25   43.0
tolls_amount  10000.0   0.623447   6.437507  -6.12   0.000   0.0   0.00  612.0
total_amount  10000.0  22.564659  19.209255 -65.92  12.375  16.3  22.88  671.8

#### Exercise 1.4
##### Isolate the `fare_amount`, `tip_amount`, `tolls_amount`, and `total_amount` for the longest trip by distance (`trip_distance`).

In [8]:
df.trip_distance.max(), max(df.trip_distance)

(38.11, 38.11)

In [9]:
df[df.trip_distance == df.trip_distance.max()]


vendorid     tpep_pickup_datetime    tpep_dropoff_datetime  \
8338         2  2019-10-23T16:50:53.000  2019-10-24T15:32:55.000   

      passenger_count  trip_distance  ratecodeid store_and_fwd_flag  \
8338                1          38.11           5                  N   

      pulocationid  dolocationid  payment_type  fare_amount  extra  mta_tax  \
8338           132           265             1        176.0    0.0      0.5   

      tip_amount  tolls_amount  improvement_surcharge  total_amount  \
8338       18.29          6.12                    0.3        201.21   

      congestion_surcharge  
8338                   0.0

In [10]:
df[df.trip_distance==max(df.trip_distance)][['fare_amount','tip_amount','tolls_amount','total_amount']]

fare_amount  tip_amount  tolls_amount  total_amount
8338        176.0       18.29          6.12        201.21

In [11]:
# conver trip_distance to kilometers
# vectorized operation
df['trip_distance_km'] = df.trip_distance * 1.60934 # this will loop through the data and convert each row
df.head(5)

vendorid     tpep_pickup_datetime    tpep_dropoff_datetime  \
0         2  2019-10-23T16:39:42.000  2019-10-23T17:14:10.000   
1         1  2019-10-23T16:32:08.000  2019-10-23T16:45:26.000   
2         2  2019-10-23T16:08:44.000  2019-10-23T16:21:11.000   
3         2  2019-10-23T16:22:44.000  2019-10-23T16:43:26.000   
4         2  2019-10-23T16:45:11.000  2019-10-23T16:58:49.000   

   passenger_count  trip_distance  ratecodeid store_and_fwd_flag  \
0                1           7.93           1                  N   
1                1           2.00           1                  N   
2                1           1.36           1                  N   
3                1           1.00           1                  N   
4                1           1.96           1                  N   

   pulocationid  dolocationid  payment_type  fare_amount  extra  mta_tax  \
0           138           170             1         29.5    1.0      0.5   
1            11            26             1         10.5    1.0      0.5   
2           163           162             1          9.5    1.0      0.5   
3           170           163             1         13.0    1.0      0.5   
4           163           236             1         10.5    1.0      0.5   

   tip_amount  tolls_amount  improvement_surcharge  total_amount  \
0        7.98          6.12                    0.3         47.90   
1        0.00          0.00                    0.3         12.30   
2        2.00          0.00                    0.3         15.80   
3        4.32          0.00                    0.3         21.62   
4        0.50          0.00                    0.3         15.30   

   congestion_surcharge  trip_distance_km  
0                   2.5         12.762066  
1                   0.0          3.218680  
2                   2.5          2.188702  
3                   2.5          1.609340  
4                   2.5          3.154306

---

### Section 2

#### Exercise 2.1
##### Read in the meteorite data from the `Meteorite_Landings.csv` file, rename the `mass (g)` column to `mass`, and drop all the latitude and longitude columns. Sort the result by mass in descending order.

In [2]:
import pandas as pd
meteorites = pd.read_csv("../data/Meteorite_Landings.csv")
meteorites.columns

Index(['name', 'id', 'nametype', 'recclass', 'mass (g)', 'fall', 'year',
       'reclat', 'reclong', 'GeoLocation'],
      dtype='object')

In [3]:
meteorites.rename(columns={'mass (g)':"mass"}, inplace=True)
meteorites.head()

name   id nametype     recclass      mass  fall  \
0    Aachen    1    Valid           L5      21.0  Fell   
1    Aarhus    2    Valid           H6     720.0  Fell   
2      Abee    6    Valid          EH4  107000.0  Fell   
3  Acapulco   10    Valid  Acapulcoite    1914.0  Fell   
4   Achiras  370    Valid           L6     780.0  Fell   

                     year    reclat    reclong           GeoLocation  
0  01/01/1880 12:00:00 AM  50.77500    6.08333     (50.775, 6.08333)  
1  01/01/1951 12:00:00 AM  56.18333   10.23333  (56.18333, 10.23333)  
2  01/01/1952 12:00:00 AM  54.21667 -113.00000    (54.21667, -113.0)  
3  01/01/1976 12:00:00 AM  16.88333  -99.90000     (16.88333, -99.9)  
4  01/01/1902 12:00:00 AM -33.16667  -64.95000   (-33.16667, -64.95)

In [4]:
# lets drop last 3 columns
# this works when you know exatly which columns you want to drop
# meteorites.drop(columns=['GeoLocation', 'reclat', 'reclong'], inplace=True)
# so I select all columns except the last 3
meteorites = meteorites[meteorites.columns[:-3]]
meteorites.head()

name   id nametype     recclass      mass  fall                    year
0    Aachen    1    Valid           L5      21.0  Fell  01/01/1880 12:00:00 AM
1    Aarhus    2    Valid           H6     720.0  Fell  01/01/1951 12:00:00 AM
2      Abee    6    Valid          EH4  107000.0  Fell  01/01/1952 12:00:00 AM
3  Acapulco   10    Valid  Acapulcoite    1914.0  Fell  01/01/1976 12:00:00 AM
4   Achiras  370    Valid           L6     780.0  Fell  01/01/1902 12:00:00 AM

In [5]:
meteorites.sort_values(by='mass', ascending=False, inplace=True)
meteorites.head(10)

name     id nametype         recclass        mass   fall  \
16392             Hoba  11890    Valid        Iron, IVB  60000000.0  Found   
5373         Cape York   5262    Valid      Iron, IIIAB  58200000.0  Found   
5365   Campo del Cielo   5247    Valid     Iron, IAB-MG  50000000.0  Found   
5370     Canyon Diablo   5257    Valid     Iron, IAB-MG  30000000.0  Found   
3455           Armanty   2335    Valid       Iron, IIIE  28000000.0  Found   
12613           Gibeon  10912    Valid        Iron, IVA  26000000.0  Found   
5468        Chupaderos   5363    Valid      Iron, IIIAB  24300000.0  Found   
26297      Mundrabilla  16852    Valid    Iron, IAB-ung  24000000.0  Found   
920       Sikhote-Alin  23593    Valid       Iron, IIAB  23000000.0   Fell   
5016        Bacubirito   4919    Valid  Iron, ungrouped  22000000.0  Found   

                         year  
16392  01/01/1920 12:00:00 AM  
5373   01/01/1818 12:00:00 AM  
5365   12/22/1575 12:00:00 AM  
5370   01/01/1891 12:00:00 AM  
3455   01/01/1898 12:00:00 AM  
12613  01/01/1836 12:00:00 AM  
5468   01/01/1852 12:00:00 AM  
26297  01/01/1911 12:00:00 AM  
920    01/01/1947 12:00:00 AM  
5016   01/01/1863 12:00:00 AM

In [6]:
meteorites['kg'] = meteorites.mass / 1000
meteorites.head(10)

name     id nametype         recclass        mass   fall  \
16392             Hoba  11890    Valid        Iron, IVB  60000000.0  Found   
5373         Cape York   5262    Valid      Iron, IIIAB  58200000.0  Found   
5365   Campo del Cielo   5247    Valid     Iron, IAB-MG  50000000.0  Found   
5370     Canyon Diablo   5257    Valid     Iron, IAB-MG  30000000.0  Found   
3455           Armanty   2335    Valid       Iron, IIIE  28000000.0  Found   
12613           Gibeon  10912    Valid        Iron, IVA  26000000.0  Found   
5468        Chupaderos   5363    Valid      Iron, IIIAB  24300000.0  Found   
26297      Mundrabilla  16852    Valid    Iron, IAB-ung  24000000.0  Found   
920       Sikhote-Alin  23593    Valid       Iron, IIAB  23000000.0   Fell   
5016        Bacubirito   4919    Valid  Iron, ungrouped  22000000.0  Found   

                         year       kg  
16392  01/01/1920 12:00:00 AM  60000.0  
5373   01/01/1818 12:00:00 AM  58200.0  
5365   12/22/1575 12:00:00 AM  50000.0  
5370   01/01/1891 12:00:00 AM  30000.0  
3455   01/01/1898 12:00:00 AM  28000.0  
12613  01/01/1836 12:00:00 AM  26000.0  
5468   01/01/1852 12:00:00 AM  24300.0  
26297  01/01/1911 12:00:00 AM  24000.0  
920    01/01/1947 12:00:00 AM  23000.0  
5016   01/01/1863 12:00:00 AM  22000.0

#### Exercise 2.2
##### Using the meteorite data from the `Meteorite_Landings.csv` file, update the `year` column to only contain the year, convert it to a numeric data type, and create a new column indicating whether the meteorite was observed falling before 1970. Set the index to the `id` column and extract all the rows with IDs between 10,036 and 10,040 (inclusive) with `loc[]`.

###### **Hint 1**: Use `year.str.slice()` to grab a substring.

###### **Hint 2**: Make sure to sort the index before using `loc[]` to select the range.

###### **Bonus**: There's a data entry error in the `year` column. Can you find it? (Don't spend too much time on this.)

In [7]:
meteorites['year'] = pd.to_datetime(meteorites['year'],errors = 'coerce' )  # change the format to date-time beforae
meteorites.head()

name     id nametype      recclass        mass   fall  \
16392             Hoba  11890    Valid     Iron, IVB  60000000.0  Found   
5373         Cape York   5262    Valid   Iron, IIIAB  58200000.0  Found   
5365   Campo del Cielo   5247    Valid  Iron, IAB-MG  50000000.0  Found   
5370     Canyon Diablo   5257    Valid  Iron, IAB-MG  30000000.0  Found   
3455           Armanty   2335    Valid    Iron, IIIE  28000000.0  Found   

            year       kg  
16392 1920-01-01  60000.0  
5373  1818-01-01  58200.0  
5365         NaT  50000.0  
5370  1891-01-01  30000.0  
3455  1898-01-01  28000.0

In [8]:
meteorites['only_year'] = pd.DatetimeIndex(meteorites['year']).year
meteorites.head()

name     id nametype      recclass        mass   fall  \
16392             Hoba  11890    Valid     Iron, IVB  60000000.0  Found   
5373         Cape York   5262    Valid   Iron, IIIAB  58200000.0  Found   
5365   Campo del Cielo   5247    Valid  Iron, IAB-MG  50000000.0  Found   
5370     Canyon Diablo   5257    Valid  Iron, IAB-MG  30000000.0  Found   
3455           Armanty   2335    Valid    Iron, IIIE  28000000.0  Found   

            year       kg  only_year  
16392 1920-01-01  60000.0     1920.0  
5373  1818-01-01  58200.0     1818.0  
5365         NaT  50000.0        NaN  
5370  1891-01-01  30000.0     1891.0  
3455  1898-01-01  28000.0     1898.0

In [9]:
meteorites.set_index('id', inplace=True)
meteorites.head()

name nametype      recclass        mass   fall       year  \
id                                                                            
11890             Hoba    Valid     Iron, IVB  60000000.0  Found 1920-01-01   
5262         Cape York    Valid   Iron, IIIAB  58200000.0  Found 1818-01-01   
5247   Campo del Cielo    Valid  Iron, IAB-MG  50000000.0  Found        NaT   
5257     Canyon Diablo    Valid  Iron, IAB-MG  30000000.0  Found 1891-01-01   
2335           Armanty    Valid    Iron, IIIE  28000000.0  Found 1898-01-01   

            kg  only_year  
id                         
11890  60000.0     1920.0  
5262   58200.0     1818.0  
5247   50000.0        NaN  
5257   30000.0     1891.0  
2335   28000.0     1898.0

In [12]:
meteorites['sub_str'] = meteorites.fall.str[:4]
meteorites.head()


name nametype      recclass        mass   fall       year  \
id                                                                            
11890             Hoba    Valid     Iron, IVB  60000000.0  Found 1920-01-01   
5262         Cape York    Valid   Iron, IIIAB  58200000.0  Found 1818-01-01   
5247   Campo del Cielo    Valid  Iron, IAB-MG  50000000.0  Found        NaT   
5257     Canyon Diablo    Valid  Iron, IAB-MG  30000000.0  Found 1891-01-01   
2335           Armanty    Valid    Iron, IIIE  28000000.0  Found 1898-01-01   

            kg  only_year sub_str  
id                                 
11890  60000.0     1920.0    Foun  
5262   58200.0     1818.0    Foun  
5247   50000.0        NaN    Foun  
5257   30000.0     1891.0    Foun  
2335   28000.0     1898.0    Foun

In [15]:
# if we have strings(object type) in our data 
# we can use str slicing to get the first 4 characters
meteorites["also_year"] = meteorites.year_str.str[:4]
meteorites.head()

name nametype      recclass        mass   fall       year  \
id                                                                            
11890             Hoba    Valid     Iron, IVB  60000000.0  Found 1920-01-01   
5262         Cape York    Valid   Iron, IIIAB  58200000.0  Found 1818-01-01   
5247   Campo del Cielo    Valid  Iron, IAB-MG  50000000.0  Found        NaT   
5257     Canyon Diablo    Valid  Iron, IAB-MG  30000000.0  Found 1891-01-01   
2335           Armanty    Valid    Iron, IIIE  28000000.0  Found 1898-01-01   

            kg  only_year sub_str             year_str also_year  
id                                                                
11890  60000.0     1920.0    Foun  1920-01-01 00:00:00      1920  
5262   58200.0     1818.0    Foun  1818-01-01 00:00:00      1818  
5247   50000.0        NaN    Foun                  NaT       NaT  
5257   30000.0     1891.0    Foun  1891-01-01 00:00:00      1891  
2335   28000.0     1898.0    Foun  1898-01-01 00:00:00      1898

In [13]:
meteorites.fall.unique()

array(['Found', 'Fell'], dtype=object)

In [14]:
meteorites["year_str"] = meteorites.year.astype(str)
meteorites.head()

name nametype      recclass        mass   fall       year  \
id                                                                            
11890             Hoba    Valid     Iron, IVB  60000000.0  Found 1920-01-01   
5262         Cape York    Valid   Iron, IIIAB  58200000.0  Found 1818-01-01   
5247   Campo del Cielo    Valid  Iron, IAB-MG  50000000.0  Found        NaT   
5257     Canyon Diablo    Valid  Iron, IAB-MG  30000000.0  Found 1891-01-01   
2335           Armanty    Valid    Iron, IIIE  28000000.0  Found 1898-01-01   

            kg  only_year sub_str             year_str  
id                                                      
11890  60000.0     1920.0    Foun  1920-01-01 00:00:00  
5262   58200.0     1818.0    Foun  1818-01-01 00:00:00  
5247   50000.0        NaN    Foun                  NaT  
5257   30000.0     1891.0    Foun  1891-01-01 00:00:00  
2335   28000.0     1898.0    Foun  1898-01-01 00:00:00

In [17]:
meteorites.sort_index(inplace=True)
meteorites.head()

name nametype recclass      mass   fall       year       kg  only_year  \
id                                                                             
1   Aachen    Valid       L5      21.0   Fell 1880-01-01    0.021     1880.0   
2   Aarhus    Valid       H6     720.0   Fell 1951-01-01    0.720     1951.0   
4    Abajo    Valid       H5     331.0  Found 1982-01-01    0.331     1982.0   
5   Abbott    Valid     H3-6   21100.0  Found 1951-01-01   21.100     1951.0   
6     Abee    Valid      EH4  107000.0   Fell 1952-01-01  107.000     1952.0   

   sub_str             year_str also_year  
id                                         
1     Fell  1880-01-01 00:00:00      1880  
2     Fell  1951-01-01 00:00:00      1951  
4     Foun  1982-01-01 00:00:00      1982  
5     Foun  1951-01-01 00:00:00      1951  
6     Fell  1952-01-01 00:00:00      1952

In [18]:
meteorites.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 45716 entries, 1 to 57458
Data columns (total 11 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   name       45716 non-null  object        
 1   nametype   45716 non-null  object        
 2   recclass   45716 non-null  object        
 3   mass       45585 non-null  float64       
 4   fall       45716 non-null  object        
 5   year       45404 non-null  datetime64[ns]
 6   kg         45585 non-null  float64       
 7   only_year  45404 non-null  float64       
 8   sub_str    45716 non-null  object        
 9   year_str   45716 non-null  object        
 10  also_year  45716 non-null  object        
dtypes: datetime64[ns](1), float64(3), object(7)
memory usage: 4.2+ MB


In [19]:
# if we have a datetime data type then we use to_datetime before comparing
# we create a new column with boolean values
meteorites["before_1970"] = meteorites.year < pd.to_datetime('1970-01-01')
meteorites.head()

name nametype recclass      mass   fall       year       kg  only_year  \
id                                                                             
1   Aachen    Valid       L5      21.0   Fell 1880-01-01    0.021     1880.0   
2   Aarhus    Valid       H6     720.0   Fell 1951-01-01    0.720     1951.0   
4    Abajo    Valid       H5     331.0  Found 1982-01-01    0.331     1982.0   
5   Abbott    Valid     H3-6   21100.0  Found 1951-01-01   21.100     1951.0   
6     Abee    Valid      EH4  107000.0   Fell 1952-01-01  107.000     1952.0   

   sub_str             year_str also_year  before_1970  
id                                                      
1     Fell  1880-01-01 00:00:00      1880         True  
2     Fell  1951-01-01 00:00:00      1951         True  
4     Foun  1982-01-01 00:00:00      1982        False  
5     Foun  1951-01-01 00:00:00      1951         True  
6     Fell  1952-01-01 00:00:00      1952         True

In [23]:
# so loc is label index which means it includes the index name start and end
meteorites.loc[10036:10040] # so tricky case cae there is no 10040 but it would have been included if it was there

name nametype         recclass      mass   fall       year  \
id                                                                       
10036     Enigma    Valid               H4      94.0  Found 1967-01-01   
10037       Enon    Valid  Iron, ungrouped     763.0  Found 1883-01-01   
10038      Enshi    Valid               H5    8000.0   Fell 1974-01-01   
10039  Ensisheim    Valid              LL6  127000.0   Fell        NaT   

            kg  only_year sub_str             year_str also_year  before_1970  
id                                                                             
10036    0.094     1967.0    Foun  1967-01-01 00:00:00      1967         True  
10037    0.763     1883.0    Foun  1883-01-01 00:00:00      1883         True  
10038    8.000     1974.0    Fell  1974-01-01 00:00:00      1974        False  
10039  127.000        NaN    Fell                  NaT       NaT        False

In [24]:
meteorites.shape

(45716, 12)

In [27]:
# pandas get index as list
meteorites.index.min(), meteorites.index.max() # so there are some missing values in this index because it is not a continues index

(1, 57458)

In [28]:
# again if you need specific numeric index then you can use iloc
meteorites.iloc[0:5] # first five rows of course 0 is optional, and 5 - row 6 is not included
# so just like using head(5) but this is more flexible

name nametype recclass      mass   fall       year       kg  only_year  \
id                                                                             
1   Aachen    Valid       L5      21.0   Fell 1880-01-01    0.021     1880.0   
2   Aarhus    Valid       H6     720.0   Fell 1951-01-01    0.720     1951.0   
4    Abajo    Valid       H5     331.0  Found 1982-01-01    0.331     1982.0   
5   Abbott    Valid     H3-6   21100.0  Found 1951-01-01   21.100     1951.0   
6     Abee    Valid      EH4  107000.0   Fell 1952-01-01  107.000     1952.0   

   sub_str             year_str also_year  before_1970  
id                                                      
1     Fell  1880-01-01 00:00:00      1880         True  
2     Fell  1951-01-01 00:00:00      1951         True  
4     Foun  1982-01-01 00:00:00      1982        False  
5     Foun  1951-01-01 00:00:00      1951         True  
6     Fell  1952-01-01 00:00:00      1952         True

In [29]:
meteorites.iloc[3:8, 2:5] # so this is the same as loc but using numeric index
# so this gets rows 4 to 8 (9th row not included)
# and columns 3 to 5 (6th column not included)
# just like any other slicing in python
# 0 based index
# last value is not included

recclass      mass   fall
id                          
5      H3-6   21100.0  Found
6       EH4  107000.0   Fell
7        L6    2914.0  Found
8         H       1.2  Found
9        H4    4500.0  Found

In [21]:
meteorites.index.dtype

dtype('int64')

#### Exercise 2.3
##### Using the meteorite data from the `Meteorite_Landings.csv` file, create a pivot table that shows both the number of meteorites and the 95th percentile of meteorite mass for those that were found versus observed falling per year from 2005 through 2009 (inclusive). Hint: Be sure to convert the `year` column to a number as we did in the previous exercise.

#### Exercise 2.4
##### Using the meteorite data from the `Meteorite_Landings.csv` file, compare summary statistics of the mass column for the meteorites that were found versus observed falling.

#### Exercise 2.5
##### Using the taxi trip data in the `2019_Yellow_Taxi_Trip_Data.csv` file, resample the data to an hourly frequency based on the dropoff time. Calculate the total `trip_distance`, `fare_amount`, `tolls_amount`, and `tip_amount`, then find the 5 hours with the most tips.

--- 

### Section 3

#### Exercise 3.1
##### Using the TSA traveler throughput data in the `tsa_melted_holiday_travel.csv` file, create box plots for traveler throughput for each year in the data. Hint: Pass `kind='box'` into the `plot()` method to generate box plots.

#### Exercise 3.2
##### Using the TSA traveler throughput data in the `tsa_melted_holiday_travel.csv` file, create a heatmap that shows the 2019 TSA median traveler throughput by day of week and month.

#### Exercise 3.3
##### Annotate the medians in the box plot from *[Exercise 3.1](#Exercise-3.1)*. Hint: The `x` coordinates will be 1, 2, and 3 for 2019, 2020, and 2021, respectively. Alternatively, to avoid hardcoding values, you can use the `Axes.get_xticklabels()` method, in which case you should look at the [documentation](https://matplotlib.org/stable/api/text_api.html) for the `Text` class.